Sources/ProcedureKit/ProcedureFuture.swift

```diff
@@ -0,0 +1,334 @@
+            assert(!_didComplete, "Called complete() more than once on a ProcedurePromise.")
+            _didComplete = true
+        }
+        #endif
+        future.complete()
+    }
+}
+
+/**
+ A ProcedureFuture can be used to schedule a following block of code after it is signaled (in the future).
+ 
+ You cannot create a ProcedureFuture directly. Instead, create a ProcedurePromise and retrieve the future
+ from the promise:
+ 
+ ```swift
+ let promise = ProcedurePromise()
+ let future = promise.future // retrieve the future from the promise
+ ```
+ 
+ then after your asynchronous task is complete, complete the promise like so:
+ 
+ ```swift
+ promise.complete()
+ ```
+ 
+ and anything scheduled using the ProcedureFuture will occur after the promise has been completed.
+
+ A simple example is:
+ 
+ ```swift
+ func myAsyncFunction() -> ProcedureFuture {
+    let promise = ProcedurePromise()
+    // dispatch an asynchronous task
+    DispatchQueue.global().async {
+        // do something
+        // then complete the promise when done
+        promise.complete()
+    }
+    return promise.future
+ }
+ 
+ myAsyncFunction().then(on: DispatchQueue.global()) {
+    // execute this block after myAsyncFunction's promise completes, on DispatchQueue.global()
+ }
+ ```
+ 
+ A ProcedureFuture/ProcedurePromise does not return a value - it merely allows you to execute 
+ a block when a promise is complete.
+ 
+ To return a future value, use ProcedureResultPromise (and ProcedureResultFuture).
+*/
+public class ProcedureFuture {
+    internal let group: DispatchGroup
+
+    internal init(group: DispatchGroup = DispatchGroup()) {
+        self.group = group
+        group.enter()
+    }
+
+    /// Schedules a block for execution after the ProcedureFuture completes.
+    ///
+    /// - Parameters:
+    ///   - on: a queue [DispatchQueue, EventQueue, Procedure (which schedules it on its internal event queue)]
+    ///   - block: The block to be scheduled after the ProcedureFuture completes.
+    public func then(on eventQueueProvider: QueueProvider, block: @escaping (Void) -> Void) {
+
+        let eventQueue = eventQueueProvider.providedQueue
+        eventQueue.dispatchNotify(withGroup: group, block: {
+            block()
+        })
+    }
+}
+
+// Used by ProcedurePromise
+fileprivate extension ProcedureFuture {
+
+    fileprivate func complete() {
+        group.leave()
+    }
+}
+
+// Used by ProcedureFutureGroup
+fileprivate extension ProcedureFuture {
+    // If the future is immediately available, execute the block synchronously on the current thread
+    // otherwise, queue a future dispatch onto the QueueProvider's queue.
+    @discardableResult
+    func thenOnSelfOrLater(on eventQueueProvider: QueueProvider, block: @escaping (Void) -> Void) -> ProcedureFuture {
+        let promise = ProcedurePromise()
+        let eventQueue = eventQueueProvider.providedQueue
+        if group.wait(timeout: .now()) == DispatchTimeoutResult.success {
```

Comment:
Shouldn't need to fully qualify the `DispatchTimeoutResult` here. 

(Let me know if pointing out these naming things is annoying.)